modules/imgproc/src/subdivision2d.cpp

```diff
@@ -642,77 +634,95 @@ void Subdiv2D::calcVoronoi()
 }
 
 
-static int
-isRightOf2( const Point2f& pt, const Point2f& org, const Point2f& diff )
-{
-    double cw_area = ((double)org.x - pt.x)*diff.y - ((double)org.y - pt.y)*diff.x;
-    return (cw_area > 0) - (cw_area < 0);
-}
-
-
 int Subdiv2D::findNearest(Point2f pt, Point2f* nearestPt)
 {
     CV_INSTRUMENT_REGION();
 
     if( !validGeometry )
         calcVoronoi();
 
-    int vertex = 0, edge = 0;
-    int loc = locate( pt, edge, vertex );
+    int vertex, edge;
+    int location = locateInternal(pt, edge, vertex);
 
-    if( loc != PTLOC_ON_EDGE && loc != PTLOC_INSIDE )
+    if (location == PTLOC_VERTEX) {
+        if (nearestPt) {
+            *nearestPt = vtx[vertex].pt;
+        }
         return vertex;
+    }
+    if (location == PTLOC_EDGE || location == PTLOC_INSIDE) {
+        vertex = !vtx[edgeOrg(edge)].isideal() ? edgeOrg(edge) : edgeDst(edge);
+    }
+    if (location == PTLOC_INSIDE) {
+        int lnext_dst = edgeDst(getEdge(edge, NEXT_AROUND_LEFT));
+        vertex = !vtx[lnext_dst].isideal() ? lnext_dst : vertex;
+    }
+    if (vtx[vertex].isideal()) {
+        // empty subdivision
+        if (nearestPt) {
+            nearestPt = NULL;
+        }
+        return 0;
+    }
 
-    vertex = 0;
+    for (;;) {
+        CV_Assert(!vtx[vertex].isideal());
 
-    Point2f start;
-    edgeOrg(edge, &start);
-    Point2f diff = pt - start;
+        int first_edge = edge = rotateEdge(vtx[vertex].firstEdge, 1);
+        bool outside_edge = false;
 
-    edge = rotateEdge(edge, 1);
+        do {
+            int edge_org = edgeOrg(edge);
+            int edge_dst = edgeDst(edge);
 
-    int i, total = (int)vtx.size();
+            bool skip_edge = false;
+            Point2f bias;
```

Comment:
Sometimes ideal points bring indeterminate expressions to CounterClockwiseEx(a,b,c) and InCircleEx(a,b,c,d), which are currently resolved internally in the predicates using assumption that the ideal points are considered relative to the origin (i.e. an edge with destination at an ideal point should have origin at (0,0)). Actually more logically rigorous solution would be to allow the CCWEx and InCircleEx predicates to return special value denoting the indeterminability, so that calling code could proceed to evaluating additional terms (i.e. non-zero origin of an edge) with the predicates.